### DISCUSSION TITLE
Cannot set-up simple mesh in C++

### DISCUSSION THREAD

**Discussion by DaveDKAF** [KratosMultiphysics/Kratos#12470] [Q&A] [answered]
Could someone pls help explain what is wrong here - I use this code snippet

           Model TMACmodel;
            ModelPart & rModelPart   = TMACmodel.CreateModelPart("Main", 2);
            
            // Add Variables
            rModelPart.AddNodalSolutionStepVariable(VELOCITY);
            rModelPart.AddNodalSolutionStepVariable(DISTANCE);
            rModelPart.AddNodalSolutionStepVariable(DISTANCE_GRADIENT);

            auto& process_info = rModelPart.GetProcessInfo();
            process_info.SetValue(DOMAIN_SIZE, 2);
            process_info.SetValue(STEP, 1);
            process_info.SetValue(NL_ITERATION_NUMBER, 1);
            
            Properties::Pointer p_elem_prop = rModelPart.CreateNewProperties(0);
 
            std::vector<Node::Pointer> p_nodes;
 
            // Set the nodes USE
            for (int i = 1; i < TNPOIN+1; i++) {
            std::cout << "Setting node number: " << i << "\n";
            auto p_node = rModelPart.CreateNewNode(i,Tx[i],Ty[i],0.0); 
            p_nodes.push_back(p_node);
            }
                                    
            // Set the elements USE
            for (int i = 0; i < TNELEM; i++) {
             std::vector<NodeType::Pointer> element_nodes ({p_nodes[IKLE1[i]], p_nodes[IKLE2[i]], p_nodes[IKLE3[i]]});
             Triangle2D3 <NodeType>::Pointer p_triangle = Kratos::make_shared<Triangle2D3 <NodeType>>( PointerVector<NodeType>{element_nodes} );
             auto p_elem = Kratos::make_intrusive<Element>(i, p_triangle, p_elem_prop);
             rModelPart.AddElement(p_elem);
            }
            
            // Initialize Elements
            const auto& r_process_info = rModelPart.GetProcessInfo();
            for (auto& r_cond : rModelPart.Conditions())
                r_cond.Initialize(r_process_info);
 

            double nh = 1.123;
             
            // Set DISTANCE and other variables ON NODES
            for (std::size_t i_node = 0; i_node < rModelPart.Nodes().size(); ++i_node) {
                auto it_node = rModelPart.Nodes().begin() + i_node;
                std::cout << "i_node: " << i_node << "\n";
                it_node->SetValue(DISTANCE, 1.0);
                it_node->SetValue(NODAL_H, nh);
                it_node->SetValue(METRIC_TENSOR_2D, ZeroVector(3));
            }

I then perform this check

 **const std::string& r_variable_name = "DISTANCE";
            KRATOS_ERROR_IF_NOT(KratosComponents<Variable<double>>::Has(r_variable_name)) << "In Variable " << r_variable_name << " is not a double variable" << std::endl;**

which results in this error:

**terminate called after throwing an instance of 'Kratos::Exception'
  what():  Error: In Variable DISTANCE is not a double variable**

**Comment by DaveDKAF** [comment]
The output of: 

std::cout << "model_part: " << rModelPart << "\n";

is:

model_part: -Main- model part
    Buffer Size : 2
    Number of tables : 0
    Number of sub model parts : 0
    Current solution step index : 0
    DOMAIN_SIZE : 2
    STEP : 1
    NL_ITERATION_NUMBER : 1

    Number of Geometries  : 0
    Mesh 0 :
        Number of Nodes       : 5
        Number of Properties  : 1
        Number of Elements    : 4
        Number of Conditions  : 0
        Number of Constraints : 0

**Comment by DaveDKAF** [comment]
Also, trying to add elements more simply with

rModelPart.CreateNewElement("Element2D3N", 1, {{1,2,3}}, p_elem_prop );

results in a segmentation fault.

**Comment by loumalouomega** [comment]
How are you running this?

**Reply by loumalouomega** [reply]
If the variable is not available is because you have not imported properly Kratos

**Reply by DaveDKAF** [reply]
I have linked Kratos with a fortran test code that calls it.  I can send the complete case a zip if that is helpful.

**Comment by DaveDKAF** [comment]
I should add I can run everything fine through the python interface so Kratos is correctly compiled.

**Comment by DaveDKAF** [comment]
[test.zip](https://github.com/user-attachments/files/15898862/test.zip)

Here is everything including the compilation script I use.  I call a cpp function from fortran and I would like to set up a mesh in the cpp code based on the nodes and connectivity data that the f90 code provides.  I would very much appreciate it if you could take a look - the key code is createmesh.cpp.  I believe I linked although Kratos libs correctly but maybe not.

**Reply by loumalouomega** [reply, marked as answer]
Looking at your code is impressive you were able to link the code to fortran and it almost worked. In addition to adding the dependencies with the path and so on, the kernel must be loaded. This is done usually with the python imports but in your case you should load the kernel in C++ manually.  See "includes/kernel.h"

**Reply by DaveDKAF** [reply]
OK thank you - I shall look into that and get back to you.

**Comment by DaveDKAF** [comment]
Hi loumalouomega - including the kernel and starting the code with this

          Model TMACmodel;
            Kernel mKernel;
            
            Kernel::GetApplicationsList().clear();
            TMACmodel.Reset();

has allowed me to progress - I can now set the nodes and elements in the convential way

// Set the nodes USE
            for (int i = 1; i < TNPOIN+1; i++) {
             r_Model_Part.CreateNewNode(i,Tx[i],Ty[i],0.0);
            }
                                    
            // Set the elements USE
            for (int i = 0; i < TNELEM; i++) {
            r_Model_Part.CreateNewElement("Element2D3N", i, {{IKLE1[i],IKLE2[i],IKLE3[i]}}, p_elem_prop);
            }

thank you very much indeed!

**Reply by loumalouomega** [reply]
![](https://media1.giphy.com/media/5xaOcLGvzHxDKjufnLW/giphy.gif?cid=6c09b952vuw8j7ahaw1e6r09b334uvwlausbbzlzl8fml230&ep=v1_internal_gif_by_id&rid=giphy.gif&ct=g)

Close discussion if you consider it is finished.

**Reply by DaveDKAF** [reply]
Have done - the linked code is now working well.
